ddprofiler/app/src/main/java/ddprofiler/preanalysis/PreAnalyzer.java

```diff
@@ -130,26 +150,100 @@ public Map<Attribute, Values> readRows(int num) {
 
     private void calculateDataTypes(Map<Attribute, List<String>> data) {
         // estimate data type for each attribute
-        for (Entry<Attribute, List<String>> e : data.entrySet()) {
-            Attribute a = e.getKey();
-            // Only if the type is not already known
-            if (!a.getColumnType().equals(AttributeType.UNKNOWN))
+        for (Entry<Attribute, List<String>> entry : data.entrySet()) {
+            Attribute attribute = entry.getKey();
+
+            // If the type is known, skip
+            if (!attribute.getColumnType().equals(AttributeType.UNKNOWN))
                 continue;
-            AttributeType aType;
+
+            AttributeType attributeType;
             if (pc.getBoolean(ProfilerConfig.EXPERIMENTAL)) {
                 // In experimental mode - force all data to be strings
-                aType = AttributeType.STRING;
+                attributeType = AttributeType.STRING;
             } else {
-                aType = typeOfValue(e.getValue());
+                attributeType = typeOfValue(entry.getValue());
             }
-            if (aType == null) {
+            if (attributeType == null) {
                 continue; // Means that data was dirty/anomaly, so skip value
             }
-            a.setColumnType(aType);
-            // a.setColumnType(AttributeType.STRING);
+            attribute.setColumnType(attributeType);
         }
     }
 
+    private void estimateSemanticTypes(Map<Attribute, List<String>> data) {
+        // estimate data type for each attribute
+        for (Entry<Attribute, List<String>> entry : data.entrySet()) {
+            Attribute attribute = entry.getKey();
+            if (attribute.getSemanticType() != null && !attribute.getSemanticType().isEmpty()) {
+                continue;
+            }
+            Map<String, String> semanticType = semanticTypeOfValue(entry.getValue());
+            attribute.setSemanticType(semanticType);
+        }
+    }
+
+    private Map<String, String> semanticTypeOfValue(List<String> values) {
+        Map<String, String> semanticTypes = new HashMap<>();
+        Map<String, Integer> temporalMatchCounts = new HashMap<>();
+        Map<String, Integer> spatialMatchCounts = new HashMap<>();
+
+        for (String value : values) {
```

Comment:
In the current implementation of `typeOfValue`, each rows will be read, so I made another function to check the first 25 (arbitrary number) non-null values. For now, I use the same regex to check for spatial/temporal. Let me know if this check should use a more general pattern.

Reply:
Actually `typeOfValue` does not read every row. The function `readRows` takes a parameter `num`. It will only read this number of rows for pre-analysis. I think `num` is defined in `ProfilerConfig.NUM_RECORD_READ`. Could you check what this value is if we follow the setup in the quickstart? 

Maybe we don't need to configure a separate number (e.g., 25) in `checkSpatialOrTemporal`.

Reply:
You're right. By default, the NUM_RECORD_READ is set to 1000, so the first call of readRows (which I thought read every row) will only read 1000 rows. I have removed the separate number in checkSpatialOrTemporal so worst case, we will match the first 1000 rows (or whatever the chunk is set to) with all possible regex.

I still think separating this functionality from the `typeOfValue` function still make sense since this functionality is different from determining primitive attribute types. If we want to improve performance (with this current design, the first 1000 rows will be checked once for primitive type, then once more for spatial/temporal type), we can change how `calculateDataType` works on each column, but I think this will make the function unnecessarily complex. Let me know if you think this should be done though!

Separate number configuration removed at 89c9a4a9a9b9e6113b0cda056a319c21b3335dcf.

Reply:
I should note though that if all data in the first NUM_RECORD_READ is null (but there may contain non-null values later on in the table), the current implementation will set ColumnSemanticType to `NONE`. Should this case be considered as well?

Reply:
Great observation. Can we drop all null values before passing column data to the pre-analyzer? You could investigate how to implement this and let's create a separate PR for this issue (would be good if you could write an issue to describe this problem)

Reply:
Great! I have created an issue #73 which we could work on after this PR.